### PR TITLE
Initial TMC2208 support

### PR DESCRIPTION
**Pull request by teemuatlut**
Add TMC2208 support. For now just the initial setup and configuration.
It uses SoftwareSerial for communication to allow for easy setup with AUX pins.
I'll add M906 support once the TMC2130 PR gets merged.

### CODE REVIEW

**Comment by teemuatlut** [comment]
There's now support for TMC2208 on
M906 Set current
M911 Report otpw status
M912 Clear otpw flag

I also learned how to use c++ templates and many of the tmc2130 specific functions will now support any drivers that implement the methods used. This will help future proof these features for any future TMC drivers (like maybe those TMC5130 ones).

I've also recently rewritten the whole TMC2130Stepper library to largely match the structure of the 2208 one. It should now be very easy to create new Trinamic driver libraries and implement support for them.

As such, this PR will now touch the TMC2130 code on many places but will not change the functionality.

**Comment by teemuatlut** [comment]
Now that the basic support is getting along, I thought next I'd add some documentation about the new GCodes to marlinfw.org. Possibly even a little writeup about the SilentStepStick differences, wiring and setup with Marlin. Thoughts?

**Comment by thinkyhead** [comment]
First place to document any added GCodes is [the RepRap wiki](http://reprap.org/wiki/gcode). And then, definitely also add them to the marlin documentation too!

**Comment by thinkyhead** [comment]
Don't forget to also add to the new config files in `example_configurations/delta/FLSUN/auto_calibrate`…

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -11609,46 +11621,43 @@ void disable_all_steppers() {
     static millis_t next_cOT = 0;
     if (ELAPSED(millis(), next_cOT)) {
       next_cOT = millis() + 5000;
-      #if ENABLED(X_IS_TMC2130)
+      #if ENABLED(X_IS_TMC2130)  || ENABLED(X_IS_TMC2208)
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: If you'll be submitting a patch, also remove the extra space on these `..||.`

File: Marlin/stepper_indirection.h [hunk 2]
```diff
@@ -84,6 +89,9 @@
     #if ENABLED(HAVE_TMC2130) && ENABLED(X_IS_TMC2130)
       extern TMC2130Stepper stepperX;
     #endif
+    #if ENABLED(HAVE_TMC2208) && ENABLED(X_IS_TMC2208)
```
Comment (thinkyhead) [inline Marlin/stepper_indirection.h]: You may use `#elif ENABLED(HAVE_TMC2208) && ENABLED(X_IS_TMC2208)` to save 1 line on these, if desired.

**Comment by thinkyhead** [comment]
Everything looks decent. I'll aim to merge this soon, but can wait if you want to add more notes to the code comments.

**Comment by teemuatlut** [comment]
Should be all good now.

**Comment by teemuatlut** [comment]
EEPROM support added. It uses the same slots as TMC2130 so I didn't increase the version count.

**Comment by oysteinkrog** [comment]
@teemuatlut: I need one pin per stepper driver for this right? I don't have enough pins, but I wonder if I could use a multiplexer?

**Comment by teemuatlut** [comment]
For reading and writing you need both RX and TX pins and a resistor. This setup is also explained in my library homepage.

For just writing you only need the TX pin. But you can also program the drivers in parallel if you don't care about setting individual parameters for each driver. This way you'd only need one pin to configure as many drivers as you like.

**Comment by teemuatlut** [comment]
But FYI I haven't yet done any safeguards for a write-only setup and I don't know how it'll work in that case. Will need to investigate and likely do some changes.

**Comment by teemuatlut** [comment]
I'll try to sort out the write-only situations so maybe not merge this just yet.

**Comment by teemuatlut** [comment]
This should be okey to merge now.

**Comment by thinkyhead** [comment]
Sorry, but some additional example configuration files were gratuitously added to Marlin at the 11th hour. You'll have to make the config changes to those new configurations too before this can be merged.

**Comment by teemuatlut** [comment]
Sure thing. But later today when I get back home.

**Comment by teemuatlut** [comment]
Updated target to bugfix-1.1.x.

Side note, the TRAMS support is coming along nicely but I'll need some new cable connectors to properly start testing and running it.

**Comment by teemuatlut** [comment]
I'm guessing target for new features will be post v1.2 release?

**Comment by alex-gh** [comment]
I don't know if it is the correct place to ask, but: 

Is it feasible to use just a single pin for communication instead of two pins since the communication is single wire and half duplex?

**Comment by teemuatlut** [comment]
Yes. But remember to change the #_SERIAL_RX_PIN pins to "-1" as this will tell Marlin that you have no read capabilities.

**Comment by alex-gh** [comment]
> Yes. But remember to change the #_SERIAL_RX_PIN pins to "-1" as this will tell Marlin that you have no read capabilities.

That is not quite what I mean.
I meant something like switching the TX pin to RX after you request data from TMC2208. This way you don't lose the read capability and don't have to provide a dedicated RX pin.
rxtx

**Comment by teemuatlut** [comment]
That's interesting. The datasheet definitely didn't mention anything about setting it up like that, but I suppose it could work. There aren't any checks to whether or not the pins are the same and limitations would come from the side of softwareSerial library.

**Comment by alex-gh** [comment]
> The datasheet definitely didn't mention anything about setting it up like that, but I suppose it could work.

True, but it also says this:

> In order to ensure a clean bus transition from the master to the slave, the TMC22xx does not
immediately send the reply to a read access, but it uses a programmable delay time after which the
first reply byte becomes sent following a read request. This delay time can be set in multiples of
eight bit times using SENDDELAY time setting (default=8 bit times) according to the needs of the
master.

So I think it should be possible. However it seems like the softwareSerial library wasn't written with a single wire half duplex application in mind.

**Comment by alexxy** [comment]
Any plans to drive 2130 motor with spi instead of step/dir?

**Comment by teemuatlut** [comment]
No, I don't think there would be any advantages and it would take a significant redesign of the Marlin step drive system (not unlike TRAMS is taking).

**Comment by alexxy** [comment]
Well on some drives it allows additional control. And changes are not that large

See bellow
https://github.com/makertum/Marlin

https://github.com/makertum/Trinamic_TMC2130

**Comment by teemuatlut** [comment]
I thought you meant XDIRECT mode that's in TMC2130.
I've already made a TMC2130 library and it has surpassed that one in Marlin. So Marlin already supports TMC2130. Or is there a specific feature you'd like me to add?

**Comment by alexxy** [comment]
Links above shows how to drive stepper with only spi + CS_ENABLE (so its without step/dir signals). That may be interesting for systems with large number of steppers (e.g. X Y Z + full color CMYKW so 8 steppers totaly)

**Comment by teemuatlut** [comment]
Only way to drive a TMC2130 without step/dir is by using XDIRECT mode. I went through the commit history and saw nothing to indicate that this would be in use. Not to talk about complete overhaul of the stepper and planner functionality in Marlin. Can you point to something more specific that ties his library to the Marlin motion system?

**Comment by kfazz** [comment]
can i use the hardware UART available on AUX1 on ramps to configure the TMC2208's instread of software serial?
I'm using a homemade digipot board to set vref on all my drivers, so i really only need UART to set extruder stepper to spreadcycle mode.

**Comment by teemuatlut** [comment]
The library supports it but this PR doesn't yet make it easy. You'll need to pass a Serial object when initiating the stepper object.
I don't think you'd be able to use the AUX1 pins anyway, since they are shared with the USB communication. You could use Serial1 on pins 18 & 19 or Serial2 on 16 & 17 (AUX4) or Serial3 on 14 & 15.

Also if you want to use the VREF, you'll need to add some configuration to the ADV section. You'd also want to max the software current limit.

**Comment by kfazz** [comment]
trying to get this going on Re-ARM
Edit: https://pastebin.com/r7e40FLM
code compiles but freezes the board.
patch applies to bigfix-2.0.x with 'tmc2208 initial support' cherry-picked.

**Comment by teemuatlut** [comment]
I, and likely @p3p too, would be interested to see how you implemented the software serial for Re-Arm. From my understanding it hasn't been done yet.

There are two while loops in the TMC library that could cause your code to hang. Both use the serial available function to read out bytes and if the supporting serial library never reports back zero, it'll happily remain within the loop.
Do you have your full code online somewhere?

You may also be interested in my [master](https://github.com/teemuatlut/Marlin/commits/master) branch that adds [hardware serial support](https://github.com/teemuatlut/Marlin/commit/29a99fea374f465c7a39bfa84a9de2a39bfdc115).

**Comment by kfazz** [comment]
i got the sw uart code working.
https://github.com/kfazz/Marlin/tree/bugfix-2.0.x-lpc_swuart
 tested 9600 and 115200,  works ok send and receive.
250000 is write only until i come up with a sub 1us delay func to get the rx timing right. 
tried hooking up tmc2208 using the following test code:
https://pastebin.com/LurFy75W
i get all zeros back.
what should replyDelay be for 115200? 
Edit2:  
Arduino gets:
Res:0 Initial GCONF:64 Res:0 IOIN:536871245 Res:0 FACTORY_CONF:13 Res:0 Final GCONF:196

**Comment by kfazz** [comment]
i now have UART3 working on i2c pins tested talking to tmc2208 on ReArm

**Comment by teemuatlut** [comment]
I'm closing this one as it is quite out dated at this point and I'll create a new TMC related PR once v2.0 is released.

To anyone who might find this PR while looking for TMC2208 support, you can find the latest development at my [master branch](https://github.com/teemuatlut/Marlin/tree/master).

**Pull request closed without merging**